# Aufgabenblatt 4

## Aufgabe 16 - Schleifen

In [2]:
def loopPower(base, numberOfLoops, search):
    alreadyFound = False
    for power in range(1, numberOfLoops):
        term = base**power
        rest = term % 1000
        print("Base: %d; Power: %d; Rest: %03d" % (base, power, rest))
        if power > base and rest == search and alreadyFound == False:
            alreadyFound = True
            print("Die Zahl '%d' ist die kleinste %der-Potenz die auf %03d endet." % (power,base,search))

numberOfLoops = 0
try:
    numberOfLoops = int(input("Anzahl Schleifendurchläufe: "))
except:
    raise Exception("Die eingegebene Zahl ist ungültig!")
        
loopPower(3, numberOfLoops, 1)
loopPower(7, numberOfLoops, 1)

Anzahl Schleifendurchläufe:  101


Base: 3; Power: 1; Rest: 003
Base: 3; Power: 2; Rest: 009
Base: 3; Power: 3; Rest: 027
Base: 3; Power: 4; Rest: 081
Base: 3; Power: 5; Rest: 243
Base: 3; Power: 6; Rest: 729
Base: 3; Power: 7; Rest: 187
Base: 3; Power: 8; Rest: 561
Base: 3; Power: 9; Rest: 683
Base: 3; Power: 10; Rest: 049
Base: 3; Power: 11; Rest: 147
Base: 3; Power: 12; Rest: 441
Base: 3; Power: 13; Rest: 323
Base: 3; Power: 14; Rest: 969
Base: 3; Power: 15; Rest: 907
Base: 3; Power: 16; Rest: 721
Base: 3; Power: 17; Rest: 163
Base: 3; Power: 18; Rest: 489
Base: 3; Power: 19; Rest: 467
Base: 3; Power: 20; Rest: 401
Base: 3; Power: 21; Rest: 203
Base: 3; Power: 22; Rest: 609
Base: 3; Power: 23; Rest: 827
Base: 3; Power: 24; Rest: 481
Base: 3; Power: 25; Rest: 443
Base: 3; Power: 26; Rest: 329
Base: 3; Power: 27; Rest: 987
Base: 3; Power: 28; Rest: 961
Base: 3; Power: 29; Rest: 883
Base: 3; Power: 30; Rest: 649
Base: 3; Power: 31; Rest: 947
Base: 3; Power: 32; Rest: 841
Base: 3; Power: 33; Rest: 523
Base: 3; Power: 34;

## Aufgabe 17 - Zufallsstrings

In [94]:
import random

lenStr = 0
possibleChars = ''

try:
    lenStr = int(input("Länge der zufällige Zeichenkette: "))
    possibleChars = input("Zeichenkette: ")
except:
    raise Exception("Fehler bei der Eingabe!")
    
generatedStr = ''
for i in range(lenStr):
    generatedStr += random.choice(possibleChars)
    
print(generatedStr)

Länge der zufällige Zeichenkette:  51
Zeichenkette:  pij


iipiijjjipjppijppppjpipjiiippjipjppjpijppjiiippipjj


## Aufgabe 18 - Verschachtelte for-Schleifen

In [3]:
for i in range(1, 10 + 1):
    for j in range(1, 15 + 1):
        for k in range(1, 20 + 1):
            product = i*j*k
            squareRoot = product**(1/2)
            if squareRoot % 1 == 0:
                print("Quadratzahl: %d <- (i: %d, j: %d, k: %d)" % (product, i, j, k))

Quadratzahl: 1 <- (i: 1, j: 1, k: 1)
Quadratzahl: 4 <- (i: 1, j: 1, k: 4)
Quadratzahl: 9 <- (i: 1, j: 1, k: 9)
Quadratzahl: 16 <- (i: 1, j: 1, k: 16)
Quadratzahl: 4 <- (i: 1, j: 2, k: 2)
Quadratzahl: 16 <- (i: 1, j: 2, k: 8)
Quadratzahl: 36 <- (i: 1, j: 2, k: 18)
Quadratzahl: 9 <- (i: 1, j: 3, k: 3)
Quadratzahl: 36 <- (i: 1, j: 3, k: 12)
Quadratzahl: 4 <- (i: 1, j: 4, k: 1)
Quadratzahl: 16 <- (i: 1, j: 4, k: 4)
Quadratzahl: 36 <- (i: 1, j: 4, k: 9)
Quadratzahl: 64 <- (i: 1, j: 4, k: 16)
Quadratzahl: 25 <- (i: 1, j: 5, k: 5)
Quadratzahl: 100 <- (i: 1, j: 5, k: 20)
Quadratzahl: 36 <- (i: 1, j: 6, k: 6)
Quadratzahl: 49 <- (i: 1, j: 7, k: 7)
Quadratzahl: 16 <- (i: 1, j: 8, k: 2)
Quadratzahl: 64 <- (i: 1, j: 8, k: 8)
Quadratzahl: 144 <- (i: 1, j: 8, k: 18)
Quadratzahl: 9 <- (i: 1, j: 9, k: 1)
Quadratzahl: 36 <- (i: 1, j: 9, k: 4)
Quadratzahl: 81 <- (i: 1, j: 9, k: 9)
Quadratzahl: 144 <- (i: 1, j: 9, k: 16)
Quadratzahl: 100 <- (i: 1, j: 10, k: 10)
Quadratzahl: 121 <- (i: 1, j: 11, k: 11)
Qua

## Aufgabe 19 - Umwandlung von for- in while-Schleife

In [8]:
j = 0
i = 1
while i < 100:
    j += i
    i += 1
    print(j, end = " ")

1 3 6 10 15 21 28 36 45 55 66 78 91 105 120 136 153 171 190 210 231 253 276 300 325 351 378 406 435 465 496 528 561 595 630 666 703 741 780 820 861 903 946 990 1035 1081 1128 1176 1225 1275 1326 1378 1431 1485 1540 1596 1653 1711 1770 1830 1891 1953 2016 2080 2145 2211 2278 2346 2415 2485 2556 2628 2701 2775 2850 2926 3003 3081 3160 3240 3321 3403 3486 3570 3655 3741 3828 3916 4005 4095 4186 4278 4371 4465 4560 4656 4753 4851 4950 

## Aufgabe 20 - (Schwierige Aufgabe) Häufigkeitstest eines Ereignisses

Ich bin mir nicht sicher, ob ich die negative Binomialverteilung richtig eingesetzt habe. Ohne den Location Parameter = 20, komme ich auf 100 für den Mittelwert, was angeblich im Vergleich zur Schleife nicht stimmt. Die Schleife tendiert zum Wert 120 als Mittelwert. Wahrscheinlich setzte ich die Verteilung falsch ein. Trotzdem habe ich was gelernt und merke das mir noch viel Wissen in der Stochastik fehlt.

Mit der Schleife habe ich folgendes gemacht. Ich habe 4 Listen erstellt, wo jede Liste Faktor 10 von der vorherigen Anzahl Stichproben entspricht. Das Resultat vom Code zeigt auf, dass desto höher die Anzahl Stichproben, desto eher tendiert der Mittelwert zur Zahl 120. Ich bin überzeugt, dass das Gesetz der grossen Zahlen hier eine wichtige Rolle spielt.

In [128]:
# Anhand Negative Binomialverteilung
import numpy as np, scipy.stats as st

n = 20
p = 1/6

mean = st.nbinom.stats(n, p, moments='m', loc=20)
print("Durchschnittlich ungefähr %d-mal" % mean)


-19.00826446280992
Durchschnittlich ungefähr 120-mal


In [7]:
# Anhand Schleife und Stichproben
import random

samples10 = []
samples100 = []
samples1000 = []
samples10000 = []

def getSample():
    matchedTimes = 0
    n = 1
    while True:
        if matchedTimes >= 20:
            break
        side = random.randint(1, 6)
        if side == 6:
            matchedTimes += 1
        n += 1
    return n

for i in range(10):
    samples10.append(getSample())
    
for i in range(100):
    samples100.append(getSample())
    
for i in range(1000):
    samples1000.append(getSample())
    
for i in range(10000):
    samples10000.append(getSample())
    
mean10 = sum(samples10) / len(samples10)
mean100 = sum(samples100) / len(samples100)
mean1000 = sum(samples1000) / len(samples1000)
mean10000 = sum(samples10000) / len(samples10000)

print("Durchschnittlich ungefähr %d-mal (Bei 10 Stichproben)" % mean10)
print("Durchschnittlich ungefähr %d-mal (Bei 100 Stichproben)" % mean100)
print("Durchschnittlich ungefähr %d-mal (Bei 1000 Stichproben)" % mean1000)
print("Durchschnittlich ungefähr %d-mal (Bei 10000 Stichproben)" % mean10000)

Durchschnittlich ungefähr 116-mal (Bei 10 Stichproben)
Durchschnittlich ungefähr 120-mal (Bei 100 Stichproben)
Durchschnittlich ungefähr 121-mal (Bei 1000 Stichproben)
Durchschnittlich ungefähr 120-mal (Bei 10000 Stichproben)
